src/config.rs

```diff
@@ -2,304 +2,624 @@
 //! specified in environment variables and/or a config file.  Those variables should be accessed
 //! from here instead of loaded again elsewhere
 
-use crate::notifications::emailer::EmailMode;
-use log::info;
-use std::env;
-use std::str::FromStr;
+use serde::{Deserialize, Serialize};
 
-lazy_static! {
+/// Top level config struct that holds other area-specific configs
+#[derive(Serialize, Deserialize, Clone)]
+pub struct Config {
+    logging: LoggingConfig,
+    api: ApiConfig,
+    database: DatabaseConfig,
+    cromwell: CromwellConfig,
+    // We don't need to specify a function for these defaults because we implemented Default for
+    // the structs
+    #[serde(default)]
+    status_manager: StatusManagerConfig,
+    #[serde(default)]
+    wdl_storage: WdlStorageConfig,
+    email: Option<EmailConfig>,
+    gcloud: Option<GCloudConfig>,
+    github: Option<GithubConfig>,
+    custom_image_build: Option<CustomImageBuildConfig>,
+    validation: ValidationConfig,
+    reporting: Option<ReportingConfig>,
+}
+
+impl Config {
+    /// Create a new instance of Config by specifying sub configs.
+    ///
+    /// # Panics
+    /// Panics if attempting to create a Config that has:
+    /// 1. A value for `github` and None for `custom_image_build`,
+    /// 2. A value for `custom_image_builds` and None for `gcloud`, or
+    /// 3. A value for `reporting` and None for `gcloud`
+    pub fn new(
+        logging: LoggingConfig,
+        api: ApiConfig,
+        database: DatabaseConfig,
+        cromwell: CromwellConfig,
+        status_manager: StatusManagerConfig,
+        wdl_storage: WdlStorageConfig,
+        email: Option<EmailConfig>,
+        gcloud: Option<GCloudConfig>,
+        github: Option<GithubConfig>,
+        custom_image_build: Option<CustomImageBuildConfig>,
+        validation: ValidationConfig,
+        reporting: Option<ReportingConfig>,
+    ) -> Self {
+        // Create new config from params
+        let new_config = Config {
+            logging,
+            api,
+            database,
+            cromwell,
+            status_manager,
+            wdl_storage,
+            email,
+            gcloud,
+            github,
+            custom_image_build,
+            validation,
+            reporting,
+        };
+        // Validate it
+        new_config.validate();
+        // Return if there are not problems
+        new_config
+    }
+    /// Validates config to make sure subconfigs that are dependent on other subconfigs have those
+    /// dependencies present.  Panics if not
+    ///
+    /// # Panics
+    /// Panics if attempting to create a Config that has:
+    /// 1. A value for `github` and None for `custom_image_build`,
+    /// 2. A value for `custom_image_builds` and None for `gcloud`, or
+    /// 3. A value for `reporting` and None for `gcloud`
+    pub fn validate(&self) {
+        if self.github.is_some() && self.custom_image_build.is_none() {
```

Comment:
I was thinking about this and I'm undecided.  The stuff I have now is kind of bare minimum.  It checks to make sure your config is valid in the sense that it has all the parts it's supposed to have, but doesn't check to see if those parts are valid.  For some of that stuff, I think you'll see the failure pretty quickly.  Like if the github config has problems, you'll see error messages pretty soon after startup because it'll immediately start querying pubsub for messages.  Some of the others won't necessarily fail right away, though.  I don't think the cromwell or gcloud ones will throw errors immediately; you might have to wait until carrot tries to use them for something.  I could look into adding that sort of validation if we want to do that.  Although that might be slightly outside the scope of this branch.

Reply:
Yeah - I was thinking about ensuring that paths for github include `github` in the URL or similar things.  I don't think we should do any validation that requires actually making a connection to an outside resource / system.